streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordDeserializer.java

```diff
@@ -17,7 +17,90 @@
 package org.apache.kafka.streams.processor.internals;
 
 import org.apache.kafka.clients.consumer.ConsumerRecord;
+import org.apache.kafka.common.record.TimestampType;
+import org.apache.kafka.streams.errors.DeserializationExceptionHandler;
+import org.apache.kafka.streams.errors.StreamsException;
+import org.apache.kafka.streams.processor.ProcessorContext;
 
-interface RecordDeserializer {
-    ConsumerRecord<Object, Object> deserialize(final ConsumerRecord<byte[], byte[]> rawRecord);
+import static java.lang.String.format;
+import static org.apache.kafka.streams.StreamsConfig.DEFAULT_DESERIALIZATION_EXCEPTION_HANDLER_CLASS_CONFIG;
+
+class RecordDeserializer {
+    private final SourceNode sourceNode;
+    private final DeserializationExceptionHandler deserializationExceptionHandler;
+
+    RecordDeserializer(final SourceNode sourceNode,
+                       final DeserializationExceptionHandler deserializationExceptionHandler) {
+        this.sourceNode = sourceNode;
+        this.deserializationExceptionHandler = deserializationExceptionHandler;
+    }
+
+    /**
+     * @throws StreamsException if a deserialization exception occurs
+     */
+    @SuppressWarnings("deprecation")
+    ConsumerRecord<Object, Object> deserialize(final ConsumerRecord<byte[], byte[]> rawRecord) {
+        final Object key;
+        try {
+            key = sourceNode.deserializeKey(rawRecord.topic(), rawRecord.headers(), rawRecord.key());
+        } catch (Exception e) {
```

Comment:
I think we should not wrap it in `deserialize`, but rather in `tryDeserailize` later. The reason is that user's customized `handler` may not expect a `StreamsException` (though in our default implementation we do not care about the exception type, it may not be always the case for user customized handlers) whereas we always wrap the expected exception in StreamsException.

What we could do, is that in `deserialize` we always pass through the original exception, but log an INFO entry for recording the topic/partition/offset of the deser-errored record.

Reply:
If we suspect the deserialization function may directly throw an exception (would it ever??), we can catch differently like what we did for timestamp extractor, i.e. in `tryDeserialize` first catch `StreamsException`, and log the context and re-throw it directly (i.e. by-pass user's handler); then catch `Exception`, and call user's handler.

Reply:
I was thinking about this a little bit. I see your point, but suggest a different approach:

It seems valuable to provide there topic/partition/offset information to the user. Thus, we could have a special type of exception we hand into the handler instead of just `Exception`. This new exception type will have the original exception as well as topic/partition/offset information. It should have according getters. This way, we "educate" the user what kind of exception to expect. This would also have the advantage, that we can distinguish between key and value deserialization error (either via a boolean flag or by using two different types of exception). Also, if we don't catch in `deserialize` but in `tryDeserialize` we loos the information if key or value deserialization failed.

If we do this, we would need to update KIP-161 accordingly. WDYT? \cc @bbejeck @dguy

Reply:
User could implement a dead letter queue and add the information there.

Reply:
I just realized, that we hand in the `rawRecord` to the handler anyway. Also, our default `LogAndFailExceptionHandler` does log all information. Thus, we don't need to log anything here.